[BE] Invalidate session on password change

MihaiDruta opened the pull request:
[BE] Invalidate session on password change
https://trello.com/c/hoFh1Wrx/73-be-invalidate-session-on-password-change

Token is now invalidated on password change.

Commit:
[BE] Invalidate session on password change
https://trello.com/c/hoFh1Wrx/73-be-invalidate-session-on-password-change

Token is now invalidated on password change.

SpartakusMd commented on src/Entity/User.php:
```diff
@@ -270,6 +276,16 @@ public function eraseCredentials(): void
     {
     }
 
+    public function getPasswordChangedAt(): ?DateTimeInterface
```
If you set a default value in construct, then it's not nullable.

SpartakusMd commented on src/Entity/User.php:
```diff
@@ -279,6 +295,8 @@ public function updatedTimestamps(): void
         $this->setUpdatedAt(new DateTime('now'));
```
`now` is the default value for date time: `new DateTime('now') === new DateTime()`

SpartakusMd commented on src/Entity/User.php:
```diff
@@ -279,6 +295,8 @@ public function updatedTimestamps(): void
         $this->setUpdatedAt(new DateTime('now'));
         if ($this->getCreatedAt() === null) {
```
Remove the `if` and it's contents as the values are set in constructor.

SpartakusMd commented on src/EventListener/JWTDecodedListener.php:
```diff
@@ -0,0 +1,37 @@
+    public function onJWTDecoded(JWTDecodedEvent $event)
+    {
+        $payload = $event->getPayload();
+        $user = $this->userRepository->findOneBy(array('username' => $payload['username']));
```
As you are searching by username, add a unique constraint in the User entity so you don't have duplicate usernames.

MihaiDruta replied:
This field is already unique in User entity. 
`/**
     * The username of User
     * @ORM\Column(type="string", unique=true)
     * @Serializer\Expose()
     * @Groups({"UserDetail", "UserList"})
     * @SWG\Property()
     */
    private $username;`

SpartakusMd commented on src/Service/UserTransformer.php:
```diff
@@ -66,6 +68,7 @@ public function registerTransform(
         $entity->setSurname($dto->surname);
         $entity->setUsername($dto->username);
         $entity->setEmail($dto->email);
+        $entity->setPasswordChangedAt(new DateTime('now'));
```
Not necessary.

SpartakusMd commented on src/Service/UserTransformer.php:
```diff
@@ -128,6 +131,7 @@ public function changePasswordTransform(
             throw $passwordDoNotMatch;
         }
         $user->setPassword($this->encoder->encodePassword($user, $dto->password));
```
Hint: inside `setPassword()` you could use `$this->setPasswordChangedAt(new DateTime('now'));` and it will always update the timestamp when the password will be changed.

Commit:
[BE] Invalidate session on password change
https://trello.com/c/hoFh1Wrx/73-be-invalidate-session-on-password-change

1.Changed assert validations for username, email, name and surname from NotNull to NotBlank
2.passwordChangedAt is now not nullable
3.Fields createdAt and passwordChangedAt are now properly generated.
